fix(lsp): handle offset encoding

rish987 opened the pull request:
@black-desk @mfussenegger @mjlbach PR alternative to #15997. I'm a bit "desperate" to fix this because #15907 breaks Lean 3 support for us, since Lean 3 actually DID use UTF-32 (see Julian/lean.nvim#206), being an experimental language server. I understand that some may want to do away with UTF-32 entirely and always assume UTF-16, but I think it's worth keeping so that it's easy to use neovim as a testing ground for new language servers without forcing them to have a translation layer to unicode indices first.

One of the central changes here is that I've added the utility wrappers `vim.lsp.util.str_utfindex` and `vim.lsp.util.str_byteindex` that let you specify an encoding. I also did a bit of refactoring to clean up special case handling where possible. Might find some time to add tests later this weekend.

mjlbach commented:
Due to the release window a fix for this is unlikely to make it into 0.6. I prefer the approach I alluded to in https://github.com/neovim/neovim/pull/15997#issuecomment-941665269, that is, handling explicitly in the calling function delegating the encoding to the correct client.

mjlbach commented on runtime/lua/vim/lsp.lua:
```diff
@@ -364,7 +365,7 @@ do
       local line_ending = format_line_ending[vim.api.nvim_buf_get_option(0, 'fileformat')]
       local incremental_change = sync.compute_diff(
-        cached_buffers[bufnr], curr_lines, firstline, lastline, new_lastline, client.offset_encoding or 'utf-16', line_ending or '\n')
+        cached_buffers[bufnr], curr_lines, firstline, lastline, new_lastline, client.offset_encoding, line_ending or '\n')
```
there's no reason to remove the fallback to utf-16, it's the default for most language servers

rish987 replied:
The fallback should still happen at a lower level.

mjlbach commented on runtime/lua/vim/lsp.lua:
```diff
@@ -1458,11 +1474,7 @@ local function adjust_start_col(lnum, line, items, encoding)
-    else
-      return vim.str_byteindex(line, min_start_char, encoding == 'utf-16')
-    end
+    return util.str_byteindex(line, min_start_char, encoding)
```
How does this handle the utf-8 case?

mjlbach replied:
It does, but I don't like wrapping an api function with a function sharing the same name

mjlbach commented on runtime/lua/vim/lsp/sync.lua:
```diff
@@ -59,31 +58,14 @@ local function byte_to_utf(line, byte, offset_encoding)
-  else
-    length = #line
-  end
+  local length = vim.lsp.util.str_utfindex(line, nil, offset_encoding)
```
Where is the handling of utf-8?

mjlbach replied:
nvm, I see what you've done. I would strongly prefer *not* eliding built-in API functions.

rish987 replied:
See the definition of this new wrapper in `util.lua`

mjlbach replied:
see my comment above :)

rish987 replied:
> I would strongly prefer not eliding built-in API functions.

I just don't think that we have a good alternative here yet if we want to avoid repeating this conditional block throughout the code.

mjlbach replied:
you can name it literally anything else.

rish987 replied:
Oh gotcha. Will do. I thought you wanted to still use the old function. Turns out I don't know what "elide" means :P

rish987 replied:
How about `lsp_utfindex` and `lsp_byteindex`?

mjlbach commented:
I generally don't like this approach of adding  a request function specifically for generating positional params. I'd prefer to explicitly handle the encoding when generating the params with the correct offset encoding per client as I suggested in the other PR.

mjlbach commented on runtime/lua/vim/lsp/diagnostic.lua:
```diff
@@ -50,12 +50,12 @@ end
 ---@private
 local function line_byte_from_position(lines, lnum, col, offset_encoding)
-  if not lines or offset_encoding == "utf-8" then
+  if not lines then
```
I don't see why you would add additional indirection through a function call here

rish987 replied:
I could revert this if you prefer. I just found it a bit cleaner since we have an LSP-specific function now.

rish987 replied:
Oh I see what you mean, yeah there's no reason to not optimize this, will fix.

mfussenegger replied:
+1 for reverting the changes in this file. Keeping the scope of the PR limited makes it easier to review.

rish987 replied:
Reverted.

mjlbach commented on runtime/lua/vim/lsp/util.lua:
```diff
@@ -150,8 +188,9 @@ end
+---@param encoding string utf-8|utf-16|utf-32|nil defaults to utf-16
 --- 1-indexed
-local function get_line_byte_from_position(bufnr, position, offset_encoding)
+local function get_line_byte_from_position(bufnr, position, encoding)
```
why change from offset encoding? that's what it is in the extension to the protocol that is being adopted (offsetEncoding)

rish987 replied:
I see, I thought I would keep it general just in case we have a need to use this in a different context. Like if there's a new protocol extension that adds a field like `length_encoding` or something?

rish987 replied:
Hmm yeah so this was a result of a search and replace so it's unlikely that here in particular it would be anything other than an offset encoding. I could go through and make these parameter names more specific if you like.

mjlbach replied:
I just, in general, would prefer we keep renaming/refactoring to a minimum if it's not on the direct path to addressing the bug, and also try to preserve naming conventions from other parts of the codebase.

mfussenegger replied:
+1, please revert unrelated changes.

rish987 replied:
Reverted.

rish987 commented:
> I'd prefer to explicitly handle the encoding when generating the params with the correct offset encoding per client as I suggested in the other PR.

I think that `buf_request` has some useful checks, perhaps this is just a matter of making a more general function that `buf_request` always calls with fixed parameters?

mjlbach commented on runtime/lua/vim/lsp.lua:
```diff
@@ -103,10 +103,11 @@ end
 
 local wait_result_reason = { [-1] = "timeout"; [-2] = "interrupted"; [-3] = "error" }
 
+local UTF8, UTF16,  UTF32 = protocol.Encodings.UTF8, protocol.Encodings.UTF16, protocol.Encodings.UTF32
```
there's a multi-space gap, we don't use in general multiple declaring/unpacking locals on the same line.

mjlbach commented on runtime/lua/vim/lsp.lua:
```diff
@@ -1301,7 +1302,8 @@ nvim_command("autocmd VimLeavePre * lua vim.lsp._vim_exit_handler()")
 ---@param bufnr (number) Buffer handle, or 0 for current.
 ---@param method (string) LSP method name
----@param params (optional, table) Parameters to send to the server
+---@param params (optional, table | funtion number -> table) Parameters to send to the server, or a function
```
I don't like that we've made params polymorphic

rish987 replied:
Me neither. What do you think about my idea above about making a more general function that _only_ takes a function?

mjlbach replied:
I think if it was something like a utility function that set up the params (and then dispatched) each request with the appropriate positional params (where params was still of the old type), that would work. Generally I want to avoid introducing more middleware like buf_request/new types of requests like pos_request.

rish987 replied:
So, you'd rather precompute the params rather than getting them between client requests, do I understand correctly? And maybe pass them in to some function as a mapping from client to params?

mjlbach commented on runtime/lua/vim/lsp.lua:
```diff
@@ -1334,7 +1336,9 @@ function lsp.buf_request(bufnr, method, params, handler)
   local client_request_ids = {}
   for_each_buffer_client(bufnr, function(client, client_id, resolved_bufnr)
-      local request_success, request_id = client.request(method, params, handler, resolved_bufnr)
+      local this_params = params
```
see above comment about the type of params

black-desk commented:
Sorry for the break. I want to fix this too, but I pause to work on it because of [this comment](https://github.com/microsoft/language-server-protocol/issues/376#issuecomment-961083252). It seems that lsp finally take this encoding issue into consideration. Should we wait for the solution from them?

rish987 commented:
@mjlbach Just reworked to replace the param-generating function with a map from client IDs to parameters, and also did the same for handlers, in a new function `buf_request_general` (there's probably a better name). Let me know if that's closer to what you had in mind.

> It seems that lsp finally take this encoding issue into consideration. Should we wait for the solution from them?

I don't think we should wait. ISTM that `offsetEncoding` is here to stay, so might as well implement this now...

mjlbach commented:
> @mjlbach Just reworked to replace the param-generating function with a map from client IDs to parameters, 

Thanks, this is exactly what I had in mind. In this case do we really need the positional request or can we just pre-generate the map for the respective clients in the helper function in buf.lua?

mjlbach commented on runtime/lua/vim/lsp.lua:
```diff
@@ -1309,50 +1315,61 @@ end
 ---  - Function which can be used to cancel all the requests. You could instead
 ---    iterate all clients and call their `cancel_request()` methods.
-function lsp.buf_request(bufnr, method, params, handler)
+function lsp.buf_request_general(bufnr, method, params, handler)
```
I don't like general, maybe `buf_request_mapping`? WDYT @mfussenegger

mfussenegger replied:
If it weren't for the breaking change I'd say rename `buf_request` to `buf_request_unsafe` and use the old name for the new signature.

mjlbach replied:
why unsafe?

mjlbach commented on runtime/lua/vim/lsp/util.lua:
```diff
@@ -1638,29 +1672,33 @@ function M.try_trim_markdown_code_blocks(lines)
+---@param window (optional, number): window handle or 0 for current, defaults to current
+---@param encoding string utf-8|utf-16|utf-32|nil defaults to utf-16
+local function make_position_param(window, encoding)
+  window = window or 0
```
Shouldn't this be "buffer" (semantically speaking) since the encodings are per buffer not per window?

rish987 replied:
The way I saw it was this function gets the cursor position from a window, so passing in a buffer could result in a mismatch.

mjlbach commented on runtime/lua/vim/lsp.lua:
```diff
@@ -371,7 +377,7 @@ do
       local line_ending = buf_get_line_ending(bufnr)
       local incremental_change = sync.compute_diff(
-        cached_buffers[bufnr], curr_lines, firstline, lastline, new_lastline, client.offset_encoding or 'utf-16', line_ending)
+        cached_buffers[bufnr], curr_lines, firstline, lastline, new_lastline, client.offset_encoding, line_ending)
```
let's keep it here and remove any fallbacks below compute_diff since compute_diff is the only thing exposed from sync.lua

rish987 replied:
I could add it back for readability, but to be clear you want to revert all the changes made to `sync.lua` in this PR?

mfussenegger commented on runtime/lua/vim/lsp/protocol.lua:
```diff
@@ -296,6 +296,12 @@ local constants = {
     -- Base kind for an organize imports source action
     SourceOrganizeImports = 'source.organizeImports';
   };
+
+  Encodings = {
+    UTF8 = 'utf-8',
+    UTF16 = 'utf-16',
+    UTF32 = 'utf-32'
+  }
 }
```
So far this table is reserved for things that are defined in the specification. Afaik `Encodings` is not part of it?

rish987 replied:
Yeah it isn't, but I think that `offsetEncoding` isn't an official part of it yet either? Do you know if there is a better place to put these constants?

rish987 commented:
> do we really need the positional request or can we just pre-generate the map for the respective clients in the helper function in buf.lua?

I'd love to make it private to `buf.lua` but it looks like we need it for `lsp.omnifunc`, which I suspect was misplaced.

mfussenegger commented on runtime/lua/vim/lsp/util.lua:
```diff
@@ -90,6 +90,44 @@ local function split_lines(value)
   return split(value, '\n', true)
 end
 
+---Alternative to vim.str_utfindex that takes an encoding.
+---@param line string line to be indexed
+---@param index number byte index (utf-8)
+---@param encoding string utf-8|utf-16|utf-32|nil defaults to utf-16
+---@return number `encoding` index of `index` in `line`
+function M.str_utfindex(line, index, encoding)
+  if encoding ~= protocol.Encodings.UTF8 then
+    local col32, col16
+    if index then
+      col32, col16 = vim.str_utfindex(line, index)
+    else
+      col32, col16 = vim.str_utfindex(line)
+    end
+    if encoding == protocol.Encodings.UTF32 then
+      return col32
+    else
+      return col16
+    end
+  else
+    return index
+  end
+end
+
+---Alternative to vim.str_byteindex that takes an encoding.
+---@param line string line to be indexed
+---@param index number UTF index
+---@param encoding string utf-8|utf-16|utf-32|nil defaults to utf-16
+---@return number byte (utf-8) index of `encoding` index `index` in `line`
+function M.str_byteindex(line, index, encoding)
+  if encoding ~= protocol.Encodings.UTF8 then
+    return vim.str_byteindex(line, index, not encoding or encoding ~= protocol.Encodings.UTF32)
+  else
+    return index
+  end
+end
+
+local str_utfindex, str_byteindex = M.str_utfindex, M.str_byteindex
```
We already have `character_offset` and `get_line_byte_from_position` which are lsp specific.
If we count the `vim.` functions we'd have a total of 6 functions for the conversions. Do we really need all of these? 

The overlapping naming makes it also kinda confusing and easy to mix them up

rish987 replied:
I don't think any existing function does what we're doing here, but I'm happy to rename it, did you have anything in mind?

mfussenegger commented on runtime/lua/vim/lsp/buf.lua:
```diff
@@ -582,8 +576,11 @@ function M.code_action(context)
   end
-  local params = util.make_range_params()
-  params.context = context
+  local params = function(client)
+    local _params = util.make_range_params(0, client.offset_encoding)
+    _params.context = context
+    return _params
+  end
   code_action_request(params)
```
Unless I'm missing something `code_action_request` (and in turn `buf_request_all`) still take regular params, not a function. This looks like it is broken now?

rish987 replied:
Thanks for catching this, should be fixed now.

mfussenegger commented on runtime/lua/vim/lsp/buf.lua:
```diff
@@ -46,6 +46,14 @@ local function request(method, params, handler)
 end
 
+local function pos_request(method, params, handler)
+  validate {
+    method = {method, 's'};
+    handler = {handler, 'f', true};
+  }
```
Why add the validation here instead of within `buf_request_position` itself?

rish987 replied:
Fixed.

mfussenegger commented on runtime/lua/vim/lsp.lua:
```diff
@@ -1407,6 +1455,26 @@ function lsp.buf_request_all(bufnr, method, params, callback)
+---@param _params (optional, table) optional table to be merged into position parameters
+---@param handler (function) see |vim.lsp.buf_request()|
+---@returns (function) see |vim.lsp.buf_request()|
+function lsp.buf_request_position(method, _params, handler)
```
I'm not really a fan that we're special casing one params variant.

Not sure what to do about it - but I feel like this isn't going in the right direction.

rish987 replied:
This should be local to `buf.lua`, but I put it here because otherwise we'd have to copy code into `lsp.omnifunc`. IMO that should also be in `buf.lua`.

mjlbach replied:
I don't think omnifunc should be in buf.lua, I think like the other vim compatible functions it should be in its own file being exposed via lsp.lua (see my recent tagfunc PR)

mfussenegger reviewed (commented):
Left a couple of comments. Overall I'm -1 on this.

It makes the API surface even more complex. This isn't exactly easy to use for LSP plugin authors who use the `buf_request` API family.

So close to the 0.6 release I'd also be warry of making a change like this.

mjlbach commented on runtime/lua/vim/lsp.lua:
```diff
@@ -1309,50 +1317,61 @@ end
       -- it up and we did the request, which should be rare.
       if request_success then
-        client_request_ids[client_id] = request_id
+        return request_id
```
note to self verify the change in return type

rish987 commented:
Two more changes:

- Renamed `lsp.buf_request_general` to `lsp.buf_request_full`, and added two extra boolean params that say whether to duplicate the params/handler per client itself. This is just so that we don't have to create two more utility functions for this.
- Deleted `lsp.buf_request_position` and replaced with `util.make_position_params_per_client` for getting the `params` argument for `lsp.buf_request_full`.

rish987 commented:
On a philosophical note related to microsoft/language-server-protocol#376, I'd argue that UTF-16 is what has muddied the waters here, not UTF-32. I don't see why we even have to consider what is literally just a file size optimization when dealing with visual character positions in a file.

rish987 commented:
After looking at `tagfunc.lua`, I'm realizing there's another issue. We have variants like `buf_request_sync` and `buf_request_all` that could also want per-client params/handlers. So I'm thinking that we make a module called `full` (or something) nested within `vim.lsp` to contain alternatives for all of these. So `vim.lsp.buf_request_full` would become `vim.lsp.full.buf_request`, and we would make `vim.lsp.full.buf_request_sync` and `vim.lsp.full.buf_request_all`. What do you think?

mjlbach commented:
I think that is going in the wrong direction.

I was talking about this with Mathias last night. In general, I think that the buf_request functions which operate on multiple clients need a complete cleanup, and I would prefer not to expand the number we already have (as this PR does).

See also https://github.com/neovim/neovim/issues/16363.

We won't have this done before 0.6 is released next week, it's going to be a breaking change which would require communicating with plugin authors.

In general, the problem is the buf_request style functions do not make much sense. Most of the time we don't want to execute handlers individually for each response. We want to either:
* choose a language server ahead of time (formatting)
* compose the results (asynchronously update the display as they come in without blocking the main thread) (imagine references from multiple language servers
* selectively fallback across language servers and operate on a single response, perhaps with some prioritization (go-to definition, anything that operates on a location response)

rish987 commented:
I agree with 100% of that. I've never actually used more than one server on the same file at a time, I'd be interested to know when that would be useful. Clearly it doesn't make sense to open multiple hover windows just because you have more than one server attached for example.

It would be nice for Lean 3 to work in 0.6, but I understand if it is impossible. Also I think `clangd` has been broken for some time now, but this has probably been hard to notice since C is mostly ASCII. Would making some of the new API here private help?

mjlbach commented:
>  I've never actually used more than one server on the same file at a time, I'd be interested to know when that would be useful. 

99% of the time it is for the JS community (or users of null-ls), neither of which apply to me or Mathias hah. For JS each framework has it's own server (so you might use tsserver, cssls, htmlls, and tailwindcss for example on the same project).

>It would be nice for Lean 3 to work in 0.6, but I understand if it is impossible

Worst case 0.7 will likely be in 6 months, and I would consider a backport to 0.6.1 even if it didn't strictly obey semver. I'm a little busy right now getting other things ready for 0.6 (and we're in the last week of what was supposed to be a feature freeze).

For lean's sake you could in the lean plugin bypass the buf_request mechanism and just directly use the raw requests with a custom keymap. I'm going to prioritize this (the buf_requests cleanups, TCP support, and bringing lspconfig into core are the three focuses of 0.7) soon, but my personal schedule is a bit tight up until the release due to the holidays/school.

rish987 commented:
> For lean's sake you could in the lean plugin bypass the buf_request mechanism and just directly use the raw requests with a custom keymap.

Thanks, I'll probably do this (so no pressure).

For this PR do you think it would be okay to:

1. Backpedal on `lsp.buf_request_full`, deleting it and restoring `lsp.buf_request` to how it worked before, and deleting `util.make_position_params_per_client`.
2. Hack fix for `buf.lua` methods to use the `offset_encoding` from the first client found for the buffer (if there is one).

I'd be happy to help out with the more fundamental solution you've suggested after 0.6.

mjlbach commented:
I'm going to let @mfussenegger make the call, but I think a fine solution would be (0.6 compatible) to make position params take the offset encoding of the first client. That should mitigate the worst of it without changing the API substantially and we can do the larger refactoring in the 0.7 cycle.

rish987 commented:
^^^ Will have something ready along these lines tomorrow. Plan to revert most of `lsp.lua` and `util.lua`.

rish987 commented on runtime/lua/vim/lsp/sync.lua:
```diff
@@ -59,31 +58,14 @@ local function byte_to_utf(line, byte, offset_encoding)
-  else
-    utf_idx = byte
-  end
+  local utf_idx = vim.lsp.util.str_utfindex_enc(line, byte, offset_encoding or vim.lsp.protocol.UTF8)
```
Wanted to preserve that UTF-8 is the default here, let me know if this was intentional?

rish987 commented:
@mfussenegger @mjlbach done, let me know what you think.

mjlbach commented on runtime/lua/vim/lsp/util.lua:
```diff
@@ -1638,41 +1672,64 @@ function M.try_trim_markdown_code_blocks(lines)
+--- Utility function for getting the encoding of the first LSP client on the given buffer.
+---@param bufnr (number) buffer handle or 0 for current, defaults to current
+---@returns (string) encoding first client if there is one, nil otherwise
+function M.get_offset_encoding(bufnr)
```
This should be a private function as we'll likely remove it (prefix with _)

mjlbach commented on runtime/lua/vim/lsp/util.lua:
```diff
@@ -1638,41 +1672,64 @@ function M.try_trim_markdown_code_blocks(lines)
-  local _
-  _, col = str_utfindex(line, col)
+
+  col = str_utfindex_enc(line, col, encoding)
```
lets stick with offset_encoding everywhere

mjlbach commented on runtime/lua/vim/lsp/util.lua:
```diff
@@ -90,6 +90,44 @@ local function split_lines(value)
+---@param index number byte index (utf-8)
+---@param encoding string utf-8|utf-16|utf-32|nil defaults to utf-16
+---@return number `encoding` index of `index` in `line`
+function M.str_utfindex_enc(line, index, encoding)
```
lets make this private (_)

mjlbach commented on runtime/lua/vim/lsp/protocol.lua:
```diff
@@ -296,6 +296,12 @@ local constants = {
     -- Base kind for an organize imports source action
     SourceOrganizeImports = 'source.organizeImports';
   };
+
```
Echoing @mfussenegger's previous comment that this should not be in protocol

mjlbach commented on runtime/lua/vim/lsp/buf.lua:
```diff
@@ -46,6 +46,14 @@ local function request(method, params, handler)
   return vim.lsp.buf_request(0, method, params, handler)
 end
 
+local pos_request = function(method, _params, handler)
```
I would just manually call offset encoding instead of making this additional helper

rish987 commented:
^ done, anything else?

black-desk commented on runtime/lua/vim/lsp/util.lua:
```diff
@@ -1638,41 +1672,64 @@ function M.try_trim_markdown_code_blocks(lines)
 
+--- Utility function for getting the encoding of the first LSP client on the given buffer.
+---@param bufnr (number) buffer handle or 0 for current, defaults to current
+---@returns (string) encoding first client if there is one, nil otherwise
```
As @justinmk commented [here](https://github.com/neovim/neovim/pull/15907#issuecomment-934443194). I think we should document this limitation and show a warning when clients have different offset encodings attaching to the same buffer.

rish987 replied:
Thanks, done.

black-desk commented on runtime/lua/vim/lsp/util.lua:
```diff
@@ -1638,41 +1672,64 @@ function M.try_trim_markdown_code_blocks(lines)
+
+  local offset_encoding
+
+  vim.lsp.for_each_buffer_client(bufnr, function(client, _)
```
I am not familiar with Lua, but I think every client should have their offset_encoding member. Why do you use `for_each_buffer_client` to go through each client, but not just check if there is a client attached to this buffer?

rish987 replied:
The idea was to get the first client's encoding, do you think I should be using a different API function for this? Or does this make more sense now with the warning message?

black-desk commented:
By the way, it seems you missing to change this line:
https://github.com/neovim/neovim/pull/16382/files#diff-c6a4160a416b22cd1fd833b8f852c4fbaf1f25adf75053e82a1025aa2c607ac9R1517

mjlbach commented:
Needs a rebase, I would also avoid changes to sync.lua. I can handle cleaning up the utf-8 handling funtions at a later date. Overall I'm now ok with this PR. I'll leave final review/comments to @mfussenegger

rish987 committed:
fix(lsp): handle offset encoding

Co-authored-by: black-desk <clx814727823@gmail.com>
Co-authored-by: Mathias Fußenegger <mfussenegger@users.noreply.github.com>

rish987 commented:
Rebased and reverted `sync.lua`.

black-desk commented on runtime/lua/vim/lsp/util.lua:
```diff
@@ -1635,41 +1667,69 @@ function M.try_trim_markdown_code_blocks(lines)
+    if not offset_encoding then
+      offset_encoding = this_offset_encoding
+    elseif offset_encoding ~= this_offset_encoding then
+      vim.notify("warning: multiple different client offset_encodings detected for buffer, this is not supported yet", vim.log.levels.WARN)
```
I am ok with this, but @justinmk said:
> ... add a check in the client-attach logic ...
https://github.com/neovim/neovim/pull/15907#issuecomment-934443194

mfussenegger commented on runtime/lua/vim/lsp/util.lua:
```diff
@@ -1635,41 +1667,69 @@ function M.try_trim_markdown_code_blocks(lines)
-  local row, col = unpack(api.nvim_win_get_cursor(0))
+---@param window (optional, number): window handle or 0 for current, defaults to current
+---@param offset_encoding string utf-8|utf-16|utf-32|nil defaults to utf-16
+local function make_position_param(window, offset_encoding)
```
What do you think of changing the default of `offset_encoding` from `utf-16` to instead call call `_get_offset_encoding` ? (Same for all other `make_xy_param(s)` functions)

This would have the advantage that the fix would propagate to plugins as well and we could omit the changes in `buf.lua`

rish987 replied:
This is a great idea, my only concern is that this makes things a bit opaque on the plugin side. That is, if you're persistently storing these parameters -- as we do for our "pins" in lean.nvim -- you don't know what encoding they came from. It would probably be okay to add the encoding as a second return value, right?

rish987 replied:
Just pushed, let me know what you think.

mfussenegger replied:
> you don't know what encoding they came from. It would probably be okay to add the encoding as a second return value, right?

I think if the caller needs to know the encoding they should set it explicitly - which is still allowed. I'd prefer it if the method didn't return it as part of the result.

rish987 replied:
This would kind of force use to either use the private function `_get_offset_encoding` or have some ugly logic for independently determining the offset encoding to pass in. I don't really see an issue if the encoding is an inherent property of the position params? IMO the protocol should even be extended to include it as its own (optional) field with the params themselves.

rish987 replied:
Ok so the issue we have is that in our plugin we're also using these position params for UI features, specifically for creating extmarks. I suppose an easier and cleaner fix would be to just have another copy of these params UTF-8 encoded for this purpose alone. So will revert the second return value.

mfussenegger commented on runtime/lua/vim/lsp/util.lua:
```diff
@@ -1635,41 +1667,69 @@ function M.try_trim_markdown_code_blocks(lines)
+
+  local offset_encoding
+
+  vim.lsp.for_each_buffer_client(bufnr, function(client, _)
+    local this_offset_encoding = client.offset_encoding or "utf-16"
+    if not offset_encoding then
+      offset_encoding = this_offset_encoding
+    elseif offset_encoding ~= this_offset_encoding then
+      vim.notify("warning: multiple different client offset_encodings detected for buffer, this is not supported yet", vim.log.levels.WARN)
+    end
+  end)
```
```suggestion
  for _, client in pairs(vim.lsp.buf_get_clients()) do
    local this_offset_encoding = client.offset_encoding or "utf-16"
    if not offset_encoding then
      offset_encoding = this_offset_encoding
    elseif offset_encoding ~= this_offset_encoding then
      vim.notify("warning: multiple different client offset_encodings detected for buffer, this is not supported yet", vim.log.levels.WARN)
    end
  end
```

Could avoid the function & closure

mfussenegger commented on runtime/lua/vim/lsp/util.lua:
```diff
@@ -1722,10 +1783,11 @@ end
 ---@see https://microsoft.github.io/language-server-protocol/specifications/specification-current/#textDocumentIdentifier
-function M.make_text_document_params()
-  return { uri = vim.uri_from_bufnr(0) }
+function M.make_text_document_params(bufnr)
```
What do you think of using `win` as parameter here as well? I think the inconsistency could otherwise trip up users and lead to bugs.

(I'd generally prefer using bufnr everywhere, but I saw that's not possible because of `make_position_param` needing to get the cursor position)

rish987 replied:
I feel like that would be more confusing personally, I see `make_position_params` as an inherently window-based method because it gets your position from the cursor, which is obviously tied to the window. This function not so much.

mfussenegger replied:
> I see make_position_params as an inherently window-based method because it gets your position from the cursor, which is obviously tied to the window. This function not so much.

I think this is leaking implementation details to some degree. E.g. if you think about how you'd explain the difference:

"These functions generate parameters related to the document you're editing in the active window"

vs.

"These functions generate parameters for the document you're editing.  `make_x` takes a bufnr because it identifies the document, `make_y` takes winnr because it also needs the cursor position"

rish987 commented on runtime/lua/vim/lsp/util.lua:
```diff
@@ -90,6 +90,44 @@ local function split_lines(value)
+function M._str_utfindex_enc(line, index, encoding)
+  if encoding ~= 'utf-8' then
+    local col32, col16
+    if index then
+      col32, col16 = vim.str_utfindex(line, index)
+    else
+      col32, col16 = vim.str_utfindex(line)
+    end
```
Hmm in practice it seems like it does not, `:lua vim.str_utfindex("asdf", nil)` results in an error for me.

mjlbach replied:
could be changed but right now the bindings use gettop so the nil will still trigger the conditional https://github.com/mjlbach/neovim/blob/68e3fbec93185f3567f60933343c38bf9e4c14e1/src/nvim/lua/stdlib.c#L178-L182

mjlbach replied:
see https://github.com/neovim/neovim/pull/16448

mfussenegger replied:
Sorry I think I wasn't very clear. I was referring to `M._str_utfindex_enc` itself not accepting a `nil` value for `index`. (If it does and should, we'd have to update the `@param` annotation instead.

(Imho it shouldn't, given the intent of the function is to convert byte index to LSP position)

rish987 replied:
Fixed the param annotation.

rish987 commented on runtime/lua/vim/lsp/util.lua:
```diff
@@ -323,8 +356,9 @@ end
+---@param offset_encoding string utf-8|utf-16|utf-32|nil defaults to utf-16
 ---@see https://microsoft.github.io/language-server-protocol/specifications/specification-current/#textEdit
-function M.apply_text_edits(text_edits, bufnr)
+function M.apply_text_edits(text_edits, bufnr, offset_encoding)
```
Should be fixed now that we default to `_get_offset_encoding`.

rish987 commented:
Let me know if there's any agreement on the remaining comments, I should be able to fix anything before Tuesday.

mjlbach commented:
fwiw this is a non-api breaking change so it’s fine if this doesn’t get into 0.6.0 because it’s backportable to 0.6.1

> On Nov 28, 2021, at 10:45 PM, Rishikesh Vaishnav ***@***.***> wrote:
> 
> 
> Let me know if there's any agreement on the remaining comments, I should be able to fix anything before Tuesday.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/neovim/neovim/pull/16382#issuecomment-981266760>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/ADFTBJUTDVKVLGRGDEJP3UTUOLZO3ANCNFSM5IOQE66A>.
>

rish987 commented:
This feels fairly unobtrusive at this point, but I could take it a step further and remove all of the new parameters if you want (if you couldn't tell, I will do pretty much anything to make sure this gets into 0.6.1 😬)

mfussenegger commented:
@rish987 can you check the linter failure? I think it's related to this PR

rish987 commented:
Yeah sorry `make_range_params()` got the position before checking the offset encoding rather then after, just fixed.
